src/EquivalenceRelation.h

```diff
@@ -0,0 +1,644 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2017 The Souffle Developers. All Rights reserved
+ * Licensed under the Universal Permissive License v 1.0 as shown at:
+ * - https://opensource.org/licenses/UPL
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+/************************************************************************
+ *
+ * @file EquivalenceRelation.h
+ *
+ * Defines a binary relation interface to be used with Souffle as a relational store.
+ * Pairs inserted into this relation implicitly store a reflexive, symmetric, and transitive relation
+ * with each other.
+ *
+ ***********************************************************************/
+
+#pragma once
+
+#include "LambdaBTree.h"
+#include "UnionFind.h"
+#include "Util.h"
+#include <algorithm>
+#include <exception>
+#include <set>
+#include <unordered_map>
+#include <utility>
+
+namespace souffle {
+template <typename TupleType>
+class EquivalenceRelation {
+    using DomainInt = typename TupleType::value_type;
```

Comment:
Would it make sense to use value_type instead of DomainInt? That would be consistent with standard containers and could possibly help with future templating magic

Reply:
Good thinking - should this be applied anywhere else?